### PR TITLE
[yt-rotate] Rotate keys

### LINKED ISSUE #1913: Youtube does not scrobble on Google Chrome, but Open Spotify scrobbles, uBlock Origin is disabled
### Steps to reproduce (for bugs)

1\.
2\.
3\.

### Steps for resolving the issue

-   [x] Reauthenticate on Last.fm
    
    
    
-   [ ] Try to use unpacked extension from latest master
    
    
-   [x] Check other issues for duplicates

### Environment

-   Version used: v2.14.0
-   Browser name and version: Google Chrome Version 72.0.3626.96

### PR DESCRIPTION
**Pull request by inverse**
Double the keys in use. What do you think? 

Resolves #1913

### CODE REVIEW

**Comment by alexesprit** [comment]
The idea is good.

But I'd use promise sequences (an example of usage is in `coverartarchive` pipeline stage), or ES `async/await` feature.

With simple `for` loop the connector can use from one up to all API keys, even if API response using 1st key is successful.

**Comment by inverse** [comment]
Good call, also I think we need to make sure that we block for each state mutation.

**Comment by inverse** [comment]
@alexesprit how does this look?

**Comment by inverse** [comment]
Just need to figure out the eslint rules to allow

**Comment by inverse** [comment]
Actually there are still issues. Any pointers?

**Comment by alexesprit** [comment]
You should add this block in ESLint config:
```js
'parserOptions': {
	'ecmaVersion': 8,
}
```

**Comment by alexesprit** [comment]
I can't use the second token due to error:
<details>
  <summary>Click to expand</summary>

```json
{
 "error": {
  "errors": [
   {
    "domain": "usageLimits",
    "reason": "accessNotConfigured",
    "message": "Access Not Configured. YouTube Data API has not been used in project 177713213466 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/youtube.googleapis.com/overview?project=177713213466 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.",
    "extendedHelp": "https://console.developers.google.com/apis/api/youtube.googleapis.com/overview?project=177713213466"
   }
  ],
  "code": 403,
  "message": "Access Not Configured. YouTube Data API has not been used in project 177713213466 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/youtube.googleapis.com/overview?project=177713213466 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry."
 }
}
```
</details>

File: src/connectors/youtube.js [hunk 1]
```diff
@@ -174,25 +177,46 @@ function getVideoCategory(videoId) {
...
+async function fetchCategoryId(videoId) {
+	let found = false;
+	for (let i = 0; i < YT_API_KEYS.length; i++) {
+		let key = YT_API_KEYS[i];
+		const url = `https://www.googleapis.com/youtube/v3/videos?part=snippet&id=${videoId}&key=${key}`;
+		try {
+			const response = await fetch(url);
 			if (!response.ok) {
-				throw new Error('Invalid response');
+				throw Error(response.statusText);
```
Comment (alexesprit) [inline src/connectors/youtube.js]: It should be `throw new Error...`.

File: src/connectors/youtube.js [hunk 2]
```diff
@@ -174,25 +177,46 @@ function getVideoCategory(videoId) {
 	if (videoId === null) {
 		return null;
 	}
-	if (!categoryCache.has(videoId)) {
-		const url = `https://www.googleapis.com/youtube/v3/videos?part=snippet&id=${videoId}&key=${YT_API_KEY}`;
-		fetch(url).then((response) => {
+	if (categoryCache.has(videoId)) {
+		fetchCategoryId(videoId).then((category) => {
+			if (category === null) {
+				throw Error(`Failed to resolve category for ${videoId}`);
```
Comment (alexesprit) [inline src/connectors/youtube.js]: Well, in this case the connector will fetch category for video id constantly. I suggest to use this algorithm:
```js
fetchCategoryId(videoId).then((category) => {
	if (category === null) {
		console.log(`Failed to resolve category for ${videoId}`);
	}
	categoryCache.set(videoId, category);
}).catch((err) => {
	console.log(err);
});
```
Add category in cache anyway, even if it's null.

File: src/connectors/youtube.js [hunk 3]
```diff
@@ -174,25 +177,46 @@ function getVideoCategory(videoId) {
...
+			console.log(err);
+		});
+	}
+
+	return categoryCache.get(videoId);
+}
+
+async function fetchCategoryId(videoId) {
+	let found = false;
+	for (let i = 0; i < YT_API_KEYS.length; i++) {
```
Comment (alexesprit) [inline src/connectors/youtube.js]: Should be replaced with `for...of` loop.

File: src/connectors/youtube.js [hunk 4]
```diff
@@ -174,25 +177,46 @@ function getVideoCategory(videoId) {
...
+				return category;
 			}
-		}).catch((err) => {
-			console.log(`Unable to get category for ${videoId}: ${err.message}`);
-		});
-		return null;
+		}
+		catch (err) {
+			console.log(`Failed fetching category for ${videoId} due to ${err}`);
+		}
```
Comment (alexesprit) [inline src/connectors/youtube.js]: It's ok, but almost all code is in this format:
```js
if (condition) {
    statement
} else {
    statement
}
```
The same thing for `try...catch` blocks.

File: src/connectors/youtube.js [hunk 5]
```diff
@@ -174,25 +177,46 @@ function getVideoCategory(videoId) {
 	if (videoId === null) {
 		return null;
 	}
-	if (!categoryCache.has(videoId)) {
-		const url = `https://www.googleapis.com/youtube/v3/videos?part=snippet&id=${videoId}&key=${YT_API_KEY}`;
-		fetch(url).then((response) => {
+	if (categoryCache.has(videoId)) {
```
Comment (alexesprit) [inline src/connectors/youtube.js]: Should be:
```js
if (!categoryCache.has(videoId)) {
```

Comment (inverse) [reply src/connectors/youtube.js]: n00b mistake

File: src/connectors/youtube.js (hunk 4 above)
Comment (inverse) [reply src/connectors/youtube.js]: Ahh formatting ;)

Comment (inverse) [reply src/connectors/youtube.js]: WOuldn't this be it? https://eslint.org/docs/rules/brace-style 1tbs which we already use?

Comment (alexesprit) [reply src/connectors/youtube.js]: It's already set up for 1TBS, but does not work for `try...catch` statements.

https://eslint.org/docs/rules/brace-style

File: src/connectors/youtube.js [hunk 6]
```diff
@@ -177,14 +177,14 @@ function getVideoCategory(videoId) {
...
+	if (!categoryCache.has(videoId)) {
 		fetchCategoryId(videoId).then((category) => {
 			if (category === null) {
-				throw Error(`Failed to resolve category for ${videoId}`);
+				console.log(`Failed to resolve category for ${videoId}`);
 			}
 			categoryCache.set(videoId, category);
-		}).catch((err) => {
-			console.log(err);
+		}).catch((error) => {
```
Comment (alexesprit) [inline src/connectors/youtube.js]: I noticed this `catch` is unused, because `fetchCategoryId` never throws any error.

**Comment by inverse** [comment]
Not sure what's up with the token, I set everything up through the panel but it's causing that issue you listed.

File: src/connectors/youtube.js (hunk 2 above)
Comment (inverse) [reply src/connectors/youtube.js]: Should we tackle this seperately?

**Comment by alexesprit** [comment]
Also, maybe it's a good idea to add error response to debug output:
```js
const response = await fetch(url);
const data = await response.json();

if (!response.ok) {
	console.log(JSON.stringify(data, null, 2));
	throw new Error(response.statusText);
}
```

**Comment by inverse** [comment]
Will that work with your example? can you call await response.json() on a non-ok request?

**Comment by alexesprit** [comment]
Yes, I tested it, and it works as intented. AFAIK, `response.ok` means 2xx response code, and vice versa.

**Comment by alexesprit** [comment]
@inverse also what do you think about storing YT API keys in one place? Now one key is in my project, and one is yours. There's an option to add members to projects, so all members can view stat of all API keys and manage them.

Unfortunately, it's not possible to make several API keys in one project, so I made 2 projects, which project contains one API key. Both keys are valid; first one exceeded its quota, though, but the second key works fine. If you're in, please tell me your Google account, and I will add you to the project team (you can email me or send it via Gitter, rather that posting email here, if you want).

File: src/connectors/youtube.js [hunk 7]
```diff
@@ -175,24 +178,42 @@ function getVideoCategory(videoId) {
...
-		return null;
+		} catch (error) {
+			console.log(`Failed fetching category with ${key}`);
+		}
 	}
-	return categoryCache.get(videoId);
 
+	if (!found) {
+		return null;
+	}
```
Comment (alexesprit) [inline src/connectors/youtube.js]: We can just return null value at the end of function, the `found` variable isn't used in the function.

**Pull request merged**